pkg/cmd/roachtest/tests/multi_store_remove.go

```diff
@@ -0,0 +1,161 @@
+	startSettings := install.MakeClusterSettings()
+	// Speed up the replicate queue.
+	startSettings.Env = append(startSettings.Env, "COCKROACH_SCAN_INTERVAL=30s")
+	c.Start(ctx, t.L(), startOpts, startSettings, c.Range(1, 3))
+
+	// Confirm that there are 6 stores live.
+	t.Status("store setup")
+	conn := c.Conn(ctx, t.L(), 2)
+	defer conn.Close()
+	var count int
+	r := conn.QueryRowContext(ctx, `SELECT count(*) FROM crdb_internal.kv_store_status;`)
+	if err := r.Scan(&count); err != nil {
+		t.Fatalf("store status: %s", err)
+	}
+	wantStores := multiStoreNodes * multiStoreStoresPerNode
+	if count != wantStores {
+		t.Fatalf("expected %d stores; got %d", wantStores, count)
+	}
+
+	// Import data.
+	t.Status("importing fixture")
+	c.Run(ctx, option.WithNodes(c.Node(1)), "./cockroach", "workload", "fixtures", "import", "bank", "--db=tinybank",
+		"--payload-bytes=100", "--ranges=1000", "--rows=100000", "--seed=1", "{pgurl:1}")
+
+	// Ensure that all stores have ranges.
+	t.Status("validating stores")
+	r = conn.QueryRowContext(ctx, `SELECT count(*) FROM crdb_internal.kv_store_status WHERE range_count = 0;`)
+	if err := r.Scan(&count); err != nil {
+		t.Fatalf("range count: %s", err)
+	}
+	if count > 0 {
+		t.Fatalf("wanted no stores without ranges; found %d", count)
+	}
+
+	// Lower the dead store detection threshold to make dead store detection
+	// faster.
+	const stmt = "SET CLUSTER SETTING server.time_until_store_dead = '30s'"
+	if _, err := conn.ExecContext(ctx, stmt); err != nil {
+		t.Fatal(err)
+	}
+
+	// Bring down node 1.
+	t.Status("removing store from n1")
+	node := c.Node(1)
+	m := c.NewMonitor(ctx, node)
+	m.ExpectDeaths(1)
+	stopOpts := option.DefaultStopOpts()
+	c.Stop(ctx, t.L(), stopOpts, node)
+
+	// Start node 1 back up without one of its stores.
+	t.Status("restarting n1")
+	startOpts.RoachprodOpts.StoreCount = multiStoreStoresPerNode - 1
+	if err := c.StartE(ctx, t.L(), startOpts, startSettings, node); err != nil {
+		t.Fatal("restarting node: %s", err)
+	}
+
+	// Wait for the store to be marked as dead.
+	t.Status("awaiting store death")
+	if err := retry.ForDuration(2*time.Minute, func() error {
+		r = conn.QueryRowContext(ctx, `SELECT count(*) FROM crdb_internal.kv_store_status;`)
+		if err := r.Scan(&count); err != nil {
+			t.Fatalf("store status: %s", err)
+		}
+		want := multiStoreNodes*multiStoreStoresPerNode - 1
+		if count == want {
+			return nil
+		}
+		return errors.Newf("waiting for %d stores; got %d", want, count)
+	}); err != nil {
+		t.Fatalf("awaiting store death: %s", err)
+	}
+
+	// Wait for up-replication.
+	// NOTE: At the time of writing, under-replicated ranges are computed using
+	// node liveness, rather than store liveness, so we instead compare the range
+	// count to the current per-store replica count to compute whether all there
+	// are still under-replicated ranges from the dead store.
+	// TODO(travers): Once #123561 is solved, re-work this.
```

Comment:
cc: @nvanbenschoten @andrewbaptist - related to the conversation last week. Let me know if you have thoughts on a better way, for now.